casper/protocols/concurrent/forkchoice.py

```diff
@@ -0,0 +1,100 @@
+"""The forkchoice module implements the estimator function a concurrent schedule"""
+
+def get_ancestors(block):
+    ancestors = set()
+    stack = [block]
+
+    while any(stack):
+        curr_block = stack.pop()
+
+        if curr_block is None:
+            continue
+
+        if curr_block not in ancestors:
+            ancestors.add(curr_block)
+            stack.extend([b for b in curr_block.estimate[0]])
+
+    return ancestors
+
+def get_scores(latest_messages):
+    scores = dict()
+
+    for validator in latest_messages:
+        ancestors = get_ancestors(latest_messages[validator])
+        for b in ancestors:
+            scores[b] = scores.get(b, 0) + validator.weight
+
+    return scores
+
+def get_outputs(blocks):
+    outputs = set()
+
+    for block in blocks:
+        outputs.update(block.estimate[2])
+
+    return outputs
+
+def update_outputs(outputs, blocks):
```

Comment:
This name is not descriptive and it's unclear to the reader what is going on here without digging in.
Maybe two methods `consume_inputs` and `collect_outputs` that return a new array rather than mutating the old one.

```python
outputs = consume_inputs(outputs, blocks)
outputs = collect_outputs(outputs, blocks)
```

Maybe should mutate. Not 100% sure.